python/hail/matrixtable.py

```diff
@@ -143,26 +335,43 @@ def aggregate(self, **named_exprs):
         :class:`.MatrixTable`
             Aggregated matrix table.
         """
+
+        assert self._row_keys is not None or self._col_keys is not None
+
         named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
 
         strs = []
 
-        base, cleanup = self._parent._process_joins(*itertools.chain(
-            [x for _, x in self._groups], named_exprs.values()))
+        fixed_fields = [f.name for f in self._parent.global_schema.fields]
+        if self._row_keys is not None:
+            fixed_fields.extend(self._row_keys.keys())
+        else:
+            fixed_fields.extend([f.name for f in self._parent.row_schema.fields])
+        if self._col_keys is not None:
+            fixed_fields.extend(self._col_keys.keys())
+        else:
+            fixed_fields.extend([f.name for f in self._parent.col_schema.fields])
+
+        base, _ = self._parent._process_joins(*named_exprs.values())
         for k, v in named_exprs.items():
-            analyze('GroupedMatrixTable.aggregate', v, self._grouped_indices,
+            if k in fixed_fields:
+                raise ExpressionException("GroupedMatrixTable.group_cols_by cannot assign duplicate field '{}'".format(k))
+            analyze('GroupedMatrixTable.aggregate', v, self._fixed_indices(),
```

Comment:
I think the _fixed_indices is wrong here. The aggregation expressions should always be entry-indexed.

Reply:
those are the non-aggregation entries, as we talked about. So in the group_rows_by case they'd be the column indices, which are valid expressions to use outside of an aggregator.

Reply:
*non-aggregation axes

Reply:
ahhh, right. Nevermind.